lib/matplotlib/backends/registry.py

```diff
@@ -20,36 +21,172 @@ class BackendRegistry:
     All use of ``BackendRegistry`` should be via the singleton instance
     ``backend_registry`` which can be imported from ``matplotlib.backends``.
 
+    Each backend has a name, a module name containing the backend code, and an
+    optional GUI framework that must be running if the backend is interactive.
+    There are three sources of backends: built-in (source code is within the
+    Matplotlib repository), explicit ``module://some.backend`` syntax (backend is
+    obtained by loading the module), or via an entry point (self-registering
+    backend in an external package).
+
     .. versionadded:: 3.9
     """
-    # Built-in backends are those which are included in the Matplotlib repo.
-    # A backend with name 'name' is located in the module
-    # f'matplotlib.backends.backend_{name.lower()}'
-
-    # The capitalized forms are needed for ipython at present; this may
-    # change for later versions.
-    _BUILTIN_INTERACTIVE = [
-        "GTK3Agg", "GTK3Cairo", "GTK4Agg", "GTK4Cairo",
-        "MacOSX",
-        "nbAgg",
-        "QtAgg", "QtCairo", "Qt5Agg", "Qt5Cairo",
-        "TkAgg", "TkCairo",
-        "WebAgg",
-        "WX", "WXAgg", "WXCairo",
-    ]
-    _BUILTIN_NOT_INTERACTIVE = [
-        "agg", "cairo", "pdf", "pgf", "ps", "svg", "template",
-    ]
-    _GUI_FRAMEWORK_TO_BACKEND_MAPPING = {
-        "qt": "qtagg",
+    # Mapping of built-in backend name to GUI framework, or "headless" for no
+    # GUI framework. Built-in backends are those which are included in the
+    # Matplotlib repo. A backend with name 'name' is located in the module
+    # f"matplotlib.backends.backend_{name.lower()}"
+    _BUILTIN_BACKEND_TO_GUI_FRAMEWORK = {
+        "gtk3agg": "gtk3",
+        "gtk3cairo": "gtk3",
+        "gtk4agg": "gtk4",
+        "gtk4cairo": "gtk4",
+        "macosx": "osx",
+        "nbagg": "nbagg",
+        "notebook": "nbagg",
+        "qtagg": "qt",
+        "qtcairo": "qt",
+        "qt5agg": "qt5",
+        "qt5cairo": "qt5",
+        "qt6agg": "qt",
+        "qt6cairo": "qt",
+        "tkagg": "tk",
+        "tkcairo": "tk",
+        "webagg": "webagg",
+        "wx": "wx",
+        "wxagg": "wx",
+        "wxcairo": "wx",
+        "agg": "headless",
+        "cairo": "headless",
+        "pdf": "headless",
+        "pgf": "headless",
+        "ps": "headless",
+        "svg": "headless",
+        "template": "headless",
+    }
+
+    # Reverse mapping of gui framework to preferred built-in backend.
+    _GUI_FRAMEWORK_TO_BACKEND = {
         "gtk3": "gtk3agg",
         "gtk4": "gtk4agg",
-        "wx": "wxagg",
-        "tk": "tkagg",
-        "macosx": "macosx",
         "headless": "agg",
+        "osx": "macosx",
```

Comment:
I changed it because IPython uses `osx` for the GUI framework and `macosx` for the backend. I thought this was fine for non-IPython Matplotlib but that isn't correct as `FigureCanvasMac.required_interactive_framework = "macosx"`. I have been lucky to get away with these two different names for the same GUI framework; non-IPython Matplotlib never looks at this code, it only needs the `required_interactive_framework` to be consistent with `cbook._get_running_interactive_framework`. But it cannot stay like this, we can't have a declaration that the macOS GUI framework is `osx` when sometimes it isn't.

It would be possible to special case the `osx` GUI framework to mean `macosx` on input, but on output we have no way of knowing if the caller prefers `osx` to `macosx`. So I think the least bad solution here is to change this back to `macosx` and be entirely consistent within Matplotlib, and I will special case the `osx` <-> `macosx` in IPython.